infrastructures/infrastructure-azure/src/main/java/org/ow2/proactive/resourcemanager/nodesource/infrastructure/AzureInfrastructure.java

```diff
@@ -0,0 +1,376 @@
+        this.vmUsername = parameters[12].toString().trim();
+        this.vmPassword = parameters[13].toString().trim();
+        this.vmPublicKey = parameters[14].toString().trim();
+        this.resourceGroup = parameters[15].toString().trim();
+        this.region = parameters[16].toString().trim();
+        this.numberOfInstances = Integer.parseInt(parameters[17].toString().trim());
+        this.numberOfNodesPerInstance = Integer.parseInt(parameters[18].toString().trim());
+        this.downloadCommand = parameters[19].toString().trim();
+        this.privateNetworkCIDR = parameters[20].toString().trim();
+        this.staticPublicIP = Boolean.parseBoolean(parameters[21].toString().trim());
+        this.additionalProperties = parameters[22].toString().trim();
+
+        connectorIaasController = new ConnectorIaasController(connectorIaasURL, INFRASTRUCTURE_TYPE);
+
+    }
+
+    private void validate(Object[] parameters) {
+        if (parameters == null || parameters.length < 23) {
+            throw new IllegalArgumentException("Invalid parameters for AzureInfrastructure creation");
+        }
+
+        if (parameters[0] == null) {
+            throw new IllegalArgumentException("Azure clientId must be specified");
+        }
+
+        if (parameters[1] == null) {
+            throw new IllegalArgumentException("Azure secret key must be specified");
+        }
+
+        if (parameters[2] == null) {
+            throw new IllegalArgumentException("Azure domain or tenantId must be specified");
+        }
+
+        if (parameters[8] == null) {
+            throw new IllegalArgumentException("The Resource manager hostname must be specified");
+        }
+
+        if (parameters[9] == null) {
+            throw new IllegalArgumentException("The connector-iaas URL must be specified");
+        }
+
+        if (parameters[10] == null) {
+            throw new IllegalArgumentException("The image id must be specified");
+        }
+
+        if (parameters[12] == null) {
+            throw new IllegalArgumentException("The virtual machine username must be specified");
+        }
+
+        if (parameters[13] == null) {
+            throw new IllegalArgumentException("The virtual machine password must be specified");
+        }
+
+        if (parameters[17] == null) {
+            throw new IllegalArgumentException("The number of instances to create must be specified");
+        }
+
+        if (parameters[18] == null) {
+            throw new IllegalArgumentException("The number of nodes per instance to deploy must be specified");
+        }
+
+        if (parameters[19] == null) {
+            throw new IllegalArgumentException("The download node.jar command must be specified");
+        }
+
+        if (parameters[22] == null) {
+            parameters[22] = "";
+        }
+    }
+
+    @Override
+    public void acquireNode() {
+
+        connectorIaasController.waitForConnectorIaasToBeUP();
```

Comment:
Good question, there is the default behaviour:

```java
    private static final int MAX_RETRIES_IN_CASE_OF_ERROR = 20;

    private static final int SLEEP_TIME_RETRIES_IN_CASE_OF_ERROR = 10000;

    public void waitForConnectorIaasToBeUP() {
        int count = 0;
        while (true) {
            try {
                restClient.getInfrastructures();
                return;
            } catch (Exception e) {
                if (++count == MAX_RETRIES_IN_CASE_OF_ERROR) {
                    logger.error(e);
                    throw e;
                } else {
                    sleepFor(SLEEP_TIME_RETRIES_IN_CASE_OF_ERROR);
                }
            }
        }
    }
```

Therefore, we wait 20*10=200 sec. before throwing an exception.
Notifying the client at this point will require WEB dev to show a popup or something during the form submission. I just added an issue as well because we are definitely missing user notifications in the whole form submission (which add infrastructure & create instances!).